### PR TITLE
feat(expansions): Changes expansions to be data driven from expansion.json

### PR DESCRIPTION
**Pull request by michaelwhelehan**
* Adds a generic custom converter for boolean flags
* Adds an expansion.json config file for expansion features
* Loads expansion data from expansion.json instead of it being hardcoded

### CODE REVIEW

File: Projects/Server/ExpansionInfo.cs [hunk 1]
```diff
@@ -141,105 +144,31 @@ public class ExpansionInfo
...
-            };
+                var expansion = expansions[i];
+                if (expansion.ClientVersion != null)
+                    Table[i] = new ExpansionInfo(
+                        i,
+                        expansion.Name,
+                        expansion.ClientVersion,
+                        expansion.FeatureFlags,
+                        expansion.CharacterListFlags,
+                        expansion.HousingFlags);
```
Comment (kamronbatman) [inline Projects/Server/ExpansionInfo.cs:162]: Let's fix the alignment, the parenthesis should go on the next line.

File: Projects/Server/Json/Converters/ClientVersionConverter.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+/*************************************************************************
+ * ModernUO                                                              *
+ * Copyright 2019-2020 - ModernUO Development Team                       *
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/ClientVersionConverter.cs:3]: Let's update this to 2021 (I admit, I need to update all of the files probably)

File: Projects/Server/Json/Converters/ClientVersionConverter.cs [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ * Email: hi@modernuo.com                                                *
+ * File: MapConverter.cs                                                 *
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/ClientVersionConverter.cs:5]: Let's update this to ClientVersionConverter.cs.
I wonder if VS has an autocomplete macro like Rider does. 🤔

File: Projects/Server/Json/Converters/ClientVersionConverterFactory.cs [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/*************************************************************************
+ * ModernUO                                                              *
+ * Copyright 2019-2020 - ModernUO Development Team                       *
+ * Email: hi@modernuo.com                                                *
+ * File: Point2DConverterFactory.cs                                      *
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/ClientVersionConverterFactory.cs:5]: Same here

File: Projects/Server/Json/Converters/FlagsConverter.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
+/*************************************************************************
+ * ModernUO                                                              *
+ * Copyright 2019-2020 - ModernUO Development Team                       *
+ * Email: hi@modernuo.com                                                *
+ * File: TimeSpanConverter.cs                                            *
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/FlagsConverter.cs:5]: Same

File: Projects/Server/Json/Converters/FlagsConverter.cs [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>. *
+ *************************************************************************/
+
+using System;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+namespace Server.Json
+{
+    public class FlagsConverter<T> : JsonConverter<T> where T : struct
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/FlagsConverter.cs:22]: I think we can do `struct, Enum`.
I am also wondering if that helps us skip the (T)(object) double casting.

File: Projects/Server/Json/Converters/FlagsConverterFactory.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+/*************************************************************************
+ * ModernUO                                                              *
+ * Copyright 2019-2020 - ModernUO Development Team                       *
+ * Email: hi@modernuo.com                                                *
+ * File: TimeSpanConverterFactory.cs                                     *
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/FlagsConverterFactory.cs:5]: Same as above

File: Projects/Server/Json/Converters/FlagsConverter.cs [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+            return (T) (object) flags;
+        }
+
+        public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
+        {
+            writer.WriteStartObject();
+            foreach (var flagName in Enum.GetNames(typeof(T)))
+            {
+                if (!flagName.StartsWith("Expansion"))
```
Comment (kamronbatman) [inline Projects/Server/Json/Converters/FlagsConverter.cs:62]: What is special about `Expansion` that we have to skip it?

File: Projects/Server/Json/Converters/FlagsConverter.cs (hunk 6 above)
Comment (kamronbatman) [reply Projects/Server/Json/Converters/FlagsConverter.cs:22]: Maybe something like this?

```cs
/*************************************************************************
 * ModernUO                                                              *
 * Copyright 2019-2021 - ModernUO Development Team                       *
 * Email: hi@modernuo.com                                                *
 * File: FlagsConverter.cs                                               *
 *                                                                       *
 * This program is free software: you can redistribute it and/or modify  *
 * it under the terms of the GNU General Public License as published by  *
 * the Free Software Foundation, either version 3 of the License, or     *
 * (at your option) any later version.                                   *
 *                                                                       *
 * You should have received a copy of the GNU General Public License     *
 * along with this program.  If not, see <http://www.gnu.org/licenses/>. *
 *************************************************************************/

using System;
using System.Runtime.CompilerServices;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Server.Json
{
    public class FlagsConverter<T> : JsonConverter<T> where T : struct, Enum
    {
        public override T Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
        {
            var flags = 0ul;
            var underlyingType = Enum.GetUnderlyingType(typeof(T));

            while (true)
            {
                reader.Read();
                if (reader.TokenType == JsonTokenType.EndObject)
                {
                    break;
                }

                if (reader.TokenType != JsonTokenType.PropertyName)
                {
                    throw new JsonException("Invalid Json structure for Flag object");
                }

                var key = reader.GetString();

                reader.Read();

                if (!reader.GetBoolean() || !Enum.TryParse<T>(key, out var val))
                {
                    continue;
                }

                flags |= ConvertToUInt64(underlyingType, val);
            }

            switch (Type.GetTypeCode(underlyingType))
            {
                case TypeCode.SByte:
                    {
                        sbyte num = (sbyte)flags;
                        return Unsafe.As<sbyte, T>(ref num);
                    }
                case TypeCode.Byte:
                    {
                        byte num = (byte)flags;
                        return Unsafe.As<byte, T>(ref num);
                    }
                case TypeCode.Int16:
                    {
                        short num = (short)flags;
                        return Unsafe.As<short, T>(ref num);
                    }
                case TypeCode.UInt16:
                    {
                        ushort num = (ushort)flags;
                        return Unsafe.As<ushort, T>(ref num);
                    }
                default:
                case TypeCode.Int32:
                    {
                        int num = (int)flags;
                        return Unsafe.As<int, T>(ref num);
                    }
                case TypeCode.UInt32:
                    {
                        uint num = (uint)flags;
                        return Unsafe.As<uint, T>(ref num);
                    }
                case TypeCode.Int64:
                    {
                        long num = (long)flags;
                        return Unsafe.As<long, T>(ref num);
                    }
                case TypeCode.UInt64:
                    {
                        return Unsafe.As<ulong, T>(ref flags);
                    }
            }
        }

        private static ulong ConvertToUInt64(Type underlyingType, object value) =>
            Type.GetTypeCode(underlyingType) switch
            {
                TypeCode.SByte  => (ulong)(sbyte)value,
                TypeCode.Byte   => (byte)value,
                TypeCode.Int16  => (ulong)(short)value,
                TypeCode.UInt16 => (ushort)value,
                TypeCode.Int32  => (ulong)(int)value,
                TypeCode.UInt32 => (uint)value,
                TypeCode.Int64  => (ulong)(long)value,
                TypeCode.UInt64 => (ulong)value,
                _               => throw new InvalidOperationException()
            };

        public override void Write(Utf8JsonWriter writer, T value, JsonSerializerOptions options)
        {
            writer.WriteStartObject();
            var underlyingType = Enum.GetUnderlyingType(typeof(T));
            var intValue = ConvertToUInt64(underlyingType, value);

            foreach (var flagName in Enum.GetNames(typeof(T)))
            {
                var flag = ConvertToUInt64(underlyingType, Enum.Parse<T>(flagName, false));
                writer.WriteBoolean(flagName, (intValue & flag) != 0);
            }
            writer.WriteEndObject();
        }
    }
}
```

**Review by kamronbatman — CHANGES_REQUESTED** [review]

File: Projects/Server/Json/Converters/ClientVersionConverter.cs (hunk 3 above)
Comment (michaelwhelehan) [reply Projects/Server/Json/Converters/ClientVersionConverter.cs:5]: I use Rider haha can you show me how it works?

File: Projects/Server/Json/Converters/FlagsConverter.cs (hunk 8 above)
Comment (michaelwhelehan) [reply Projects/Server/Json/Converters/FlagsConverter.cs:62]: The expansion flags are just aggregates of the individual flags so I don't want to write those flags out to booleans

File: Projects/Server/ExpansionInfo.cs (hunk 1 above)
Comment (michaelwhelehan) [reply Projects/Server/ExpansionInfo.cs]: Done

File: Projects/Server/Json/Converters/FlagsConverter.cs (hunk 6 above)
Comment (michaelwhelehan) [reply Projects/Server/Json/Converters/FlagsConverter.cs]: Ahh thanks for providing this, I've updated it and tested that it works. Although it feels a lot more verbose than the double cast it is probably a cleaner solution in the end

File: Projects/Server/Json/Converters/ClientVersionConverter.cs (hunk 3 above)
Comment (kamronbatman) [reply Projects/Server/Json/Converters/ClientVersionConverter.cs]: @michaelwhelehan import the https://github.com/modernuo/ModernUO/blob/main/rider-settings.zip file into your settings.

Then you type `copy` and will see a dropdown for copyright, press enter, and it is done for you.

File: Projects/Server/Json/Converters/FlagsConverter.cs (hunk 8 above)
Comment (kamronbatman) [reply Projects/Server/Json/Converters/FlagsConverter.cs:121]: So let's do `Enum.GetValues(typeof(T))` to get the values in an array.
Then let's check it generically:
```cs
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        private static int GetUnderlyingTypeLength(TypeCode typeCode) =>
            typeCode switch
            {
                TypeCode.Byte   => 1,
                TypeCode.SByte  => 1,
                TypeCode.Int16  => 2,
                TypeCode.UInt16 => 2,
                TypeCode.Char   => 2,
                TypeCode.Int32  => 4,
                TypeCode.UInt32 => 4,
                TypeCode.Int64  => 8,
                TypeCode.UInt64 => 8,
                _               => 8
            };
```

```cs
var size = GetUnderlyingTypeLength(underlyingType) - 1;
...
var flag = ConvertToUInt64(underlyingType, flagValues[i]);
if (flag == 0 || (flag & size) == 0)
{
  // write value
}
```

File: Projects/Server/ExpansionInfo.cs [hunk 9]
```diff
@@ -331,4 +263,23 @@ public static ExpansionInfo GetInfo(int ex)
 
         public override string ToString() => Name;
     }
+
+    public record ExpansionConfig
+    {
+        public string Name { get; init; }
+
+        [JsonConverter(typeof(ClientVersionConverter))]
```
Comment (kamronbatman) [inline Projects/Server/ExpansionInfo.cs:271]: Client version can go into options. Only enum flags should be a custom attributed converter.

File: Projects/Server/ExpansionInfo.cs [hunk 10]
```diff
@@ -141,105 +145,33 @@ public class ExpansionInfo
...
-                    11,
-                    "Endless Journey",
-                    new ClientVersion("7.0.61.0"),
-                    FeatureFlags.ExpansionEJ,
-                    CharacterListFlags.ExpansionEJ,
-                    HousingFlags.HousingEJ
-                )
-            };
+                var expansion = expansions[i];
+                if (expansion.ClientVersion != null)
```
Comment (kamronbatman) [inline Projects/Server/ExpansionInfo.cs:156]: Almost missed this. We should always have braces around all if statements, even single line.

**Comment by kamronbatman** [comment]
I am not sure why but the Table is empty during unit tests? We should look into this before merging.

**Review by kamronbatman — APPROVED** [review]

**Pull request merged**